akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala

```diff
@@ -0,0 +1,487 @@
+      ackProbe.expectMessage(Ack("inc-4"))
+      ackProbe.expectMessage(Ack("inc-5"))
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(3, active = true))
+    }
+
+    "discard internal stash when restarted due to thrown exception" in {
+      val c = spawn(counter(nextPid()))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+      val latch = new CountDownLatch(1)
+
+      // make first command slow to ensure that all subsequent commands are enqueued first
+      c ! Slow("slow", latch, ackProbe.ref)
+
+      (1 to 10).foreach { n ⇒
+        if (n == 3)
+          c ! IncrementThenThrow(s"inc-$n", new TestException("test"), ackProbe.ref)
+        else
+          c ! Increment(s"inc-$n", ackProbe.ref)
+      }
+
+      latch.countDown()
+      ackProbe.expectMessage(Ack("slow"))
+
+      ackProbe.expectMessage(Ack("inc-1"))
+      ackProbe.expectMessage(Ack("inc-2"))
+      ackProbe.expectNoMessage()
+
+      c ! Increment("inc-11", ackProbe.ref)
+      ackProbe.expectMessage(Ack("inc-11"))
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(4, active = true))
+    }
+
+    "preserve internal stash when persist failed" in {
+      val c = spawn(counter(PersistenceId("fail-fifth-a")))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+
+      (1 to 10).foreach { n ⇒
+        c ! Increment(s"inc-$n", ackProbe.ref)
+      }
+
+      (1 to 10).foreach { n ⇒
+        if (n != 5)
+          ackProbe.expectMessage(Ack(s"inc-$n"))
+      }
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(9, active = true))
+    }
+
+    "preserve external stash when persist failed" in {
+      val c = spawn(counter(PersistenceId("fail-fifth-b")))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+
+      c ! Increment("inc-1", ackProbe.ref)
+      ackProbe.expectMessage(Ack("inc-1"))
+
+      c ! Deactivate("deact", ackProbe.ref)
+      ackProbe.expectMessage(Ack("deact"))
+
+      c ! Increment("inc-2", ackProbe.ref)
+      c ! Increment("inc-3", ackProbe.ref) // this will fail when unstashed
+      c ! Increment("inc-4", ackProbe.ref)
+      c ! Increment("inc-5", ackProbe.ref)
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(1, active = false))
+
+      c ! Activate("act", ackProbe.ref)
+      c ! Increment("inc-6", ackProbe.ref)
+
+      ackProbe.expectMessage(Ack("act"))
+      ackProbe.expectMessage(Ack("inc-2"))
+      // inc-3 failed
+      // inc-4, inc-5 still in external stash and processed due to UnstashAll that was in progress
+      ackProbe.expectMessage(Ack("inc-4"))
+      ackProbe.expectMessage(Ack("inc-5"))
+      ackProbe.expectMessage(Ack("inc-6"))
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(5, active = true))
+    }
+  }
+
+  // FIXME test combination with PoisonPill
```

Comment:
yeah, I will add that test

I think it should drain any ongoing UnstashAll, but not wait for the external stash to become empty. The reason is that it's known that UnstashAll will complete but it's know that there will be an UnstashAll any time soon

Reply:
:+1: